shelley/chain-and-ledger/executable-spec/test/Generator/Delegation/QuickCheck.hs

```diff
@@ -22,47 +22,55 @@ import           Test.QuickCheck (Gen)
 import qualified Test.QuickCheck as QC
 import           Test.Utils
 
+import           Address (scriptToCred)
 import           BaseTypes (interval0)
 import           Coin (Coin (..))
 import           ConcreteCryptoTypes (CoreKeyPair, DCert, DPState, DState, KeyPair, KeyPairs,
-                     PState, PoolParams, VKey, VrfKeyPairs, hashKeyVRF)
+                     MultiSig, MultiSigPairs, PState, PoolParams, VKey, VrfKeyPairs, hashKeyVRF)
 import           Delegation.Certificates (pattern DCertMir, pattern DeRegKey, pattern Delegate,
                      pattern GenesisDelegate, pattern MIRCert, pattern RegKey, pattern RegPool,
                      pattern RetirePool, pattern StakeCreds, decayKey, isDeRegKey)
 import           Examples (unsafeMkUnitInterval)
 import           Generator.Core.QuickCheck (genCoinList, genInteger, genWord64, toCred)
-import           Keys (GenDelegs (..), hashKey, vKey)
+import           Keys (GenDelegs (..), hashAnyKey, hashKey, vKey)
 import           Ledger.Core (dom, range, (∈), (∉))
 import           LedgerState (dstate, keyRefund, pParams, pstate, stPools, stkCreds, _dstate,
                      _genDelegs, _pstate, _stPools, _stkCreds)
 import           PParams (PParams (..), d, eMax)
 import           Slot (EpochNo (EpochNo), SlotNo (SlotNo))
-import           TxData (Credential (KeyHashObj), pattern DCertDeleg, pattern DCertGenesis,
-                     pattern DCertPool, pattern Delegation, pattern PoolParams, RewardAcnt (..),
-                     StakePools (StakePools), _poolPubKey, _poolVrf)
+import           Tx (getKeyCombinations)
+import           TxData (pattern DCertDeleg, pattern DCertGenesis, pattern DCertPool,
+                     pattern Delegation, pattern KeyHashObj, pattern PoolParams, RewardAcnt (..),
+                     pattern StakePools, _poolPubKey, _poolVrf)
 import           UTxO (deposits)
 
+data CertCred = CoreKeyCred [CoreKeyPair]
+              | KeyCred KeyPair
+              | ScriptCred (MultiSig, MultiSig)
+              deriving (Show)
+
 -- | Generate certificates and also return the associated witnesses and
 -- deposits and refunds required.
 genDCerts
   :: KeyPairs
+  -> MultiSigPairs
   -> [CoreKeyPair]
   -> VrfKeyPairs
   -> PParams
   -> DPState
   -> SlotNo
   -> Natural
-  -> Gen (Seq DCert, [KeyPair], [CoreKeyPair], Coin, Coin)
-genDCerts keys coreKeys vrfKeys pparams dpState slot ttl = do
+  -> Gen (Seq DCert, [KeyPair], [CoreKeyPair], [MultiSig], Coin, Coin)
```

Comment:
this could become `(Seq (DCert, CertCred), Coin, Coin)` or `(Seq DCert, [CertCred], Coin, Coin)` ?